Fix inventory cache interface (#50446)
ansible/ansible: docs/docsite/rst/dev_guide/developing_plugins.rst

To access the configuration settings in your plugin, use ``self.get_option(<option_name>)``. For most plugin types, the controller pre-populates the settings. If you need to populate settings explicitly, use a ``self.set_options()`` call.

> Cache plugins should always access options with ``self.get_option(<option_name>)`` to allow inventory plugins to use them.

Reviewer: ```suggestion
```

For practical examples of action plugins,
see the source code for the `action plugins included with Ansible Core <https://github.com/ansible/ansible/tree/devel/lib/ansible/plugins/action>`_

.. _developing_cache_plugins:

Cache plugins
-------------

Cache plugins store gathered facts and data retrieved by inventory plugins.

> Cache plugins should not be imported directly in the code base. Import cache plugins using the cache_loader so ``self.set_options()`` and ``self.get_option(<option_name>)`` can be used. Not using the cache_loader would restrict options to be only be available via ``ansible.constants``. This would also break its ability to be used by an inventory plugin.

Reviewer: ```suggestion
Import cache plugins using the cache_loader so you can use ``self.set_options()`` and ``self.get_option(<option_name>)``. If you import a cache plugin directly in the code base, you can only access options via ``ansible.constants``, and you break the cache plugin's ability to be used by an inventory plugin.
```

There are two base classes for cache plugins, ``BaseCacheModule`` for database-backed caches, and ``BaseCacheFileModule`` for file-backed caches.

To create a cache plugin, start by creating a new ``CacheModule`` class with the appropriate base class. If you're creating a plugin using an ``__init__`` method you should initialize the base class with any provided args and kwargs to be compatible with inventory plugin cache options. The base class calls ``self.set_options(direct=kwargs)``. After the base class ``__init__`` method is called ``self.get_option(<option_name>)`` should be used to access cache options.

New cache plugins should take the options ``_uri``, ``_prefix``, and ``_timeout`` to be consistent with existing cache plugins.

.. code-block:: python

    from ansible.plugins.cache import BaseCacheModule

    class CacheModule(BaseCacheModule):
        def __init__(self, *args, **kwargs):
            super(CacheModule, self).__init__(*args, **kwargs)
            self._connection = self.get_option('_uri')
            self._prefix = self.get_option('_prefix')
            self._timeout = self.get_option('_timeout')

> Caches that use ``BaseCacheModule`` will need to implement the methods ``get``, ``contains``, ``keys``, ``set``, ``delete``, ``flush``, and ``copy``. The ``contains`` method should return a boolean that indicates if the key exists and has not expired. Unlike file-based caches, the ``get`` method does not raise a KeyError if the cache has expired.

Reviewer: ```suggestion
If you use the ``BaseCacheModule``, you must implement the methods ``get``, ``contains``, ``keys``, ``set``, ``delete``, ``flush``, and ``copy``. The ``contains`` method should return a boolean that indicates if the key exists and has not expired. Unlike file-based caches, the ``get`` method does not raise a KeyError if the cache has expired.
```

> Caches that use ``BaseFileCacheModule`` must implement ``_load`` and ``_dump`` methods that will be called from the base class methods ``get`` and ``set``.

Reviewer: ```suggestion
If you use the ``BaseFileCacheModule``, you must implement ``_load`` and ``_dump`` methods that will be called from the base class methods ``get`` and ``set``.
```

> If the cache plugin stores JSON, ``AnsibleJSONEncoder`` should be used in the ``_dump`` or ``set`` method  and ``AnsibleJSONDecoder`` should be used in the ``_load`` or ``get`` method.

Reviewer: ```suggestion
If your cache plugin stores JSON, use ``AnsibleJSONEncoder`` in the ``_dump`` or ``set`` method  and ``AnsibleJSONDecoder`` in the ``_load`` or ``get`` method.
```
